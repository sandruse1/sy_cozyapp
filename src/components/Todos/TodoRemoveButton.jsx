import React, { Component } from 'react'

import Button from 'cozy-ui/react/Button'
import { withTodosMutations } from 'connections/allTodos'

export class TodoRemoveButton extends Component {
  constructor(props) {
    super(props)
    // initial component state
    this.state = { isWorking: false }
  }

  // delete the related todo
  removeTodo = async () => {
    const { deleteTodo, todo } = this.props
    // display a spinner during the process
    this.setState(() => ({ isWorking: true }))
    // delete the todo in the Cozy : asynchronous
    await deleteTodo(todo)
    // remove the spinner
    this.setState(() => ({ isWorking: false }))
  }

  render() {
    const { isWorking } = this.state
    return (
      <Button
        className="todo-remove-button"
        theme="danger"
        icon="delete"
        busy={isWorking}
        disabled={isWorking}
        onClick={this.removeTodo}
        extension="narrow"
      />
    )
  }
}

// get mutations from the client to use deleteTodo
export default withTodosMutations(TodoRemoveButton)
