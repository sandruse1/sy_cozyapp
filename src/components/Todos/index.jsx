import React from 'react'

import Spinner from 'cozy-ui/react/Spinner'
import TodoAdd from './TodoAdd'
import TodosList from './TodosList'
import { withTodos } from 'connections/allTodos'

export const Todos = props => {
  const { data, fetchStatus } = props
  // cozy-client statuses
  const isLoading = fetchStatus === 'loading' || fetchStatus === 'pending'
  return (
    <div className="todos">
      {isLoading ? (
        <Spinner size="xxlarge" middle />
      ) : (
        <div>
          <TodosList todos={data} />
          <TodoAdd />
        </div>
      )}
    </div>
  )
}

// get data from the client state: data, fetchStatus
export default withTodos(Todos)
