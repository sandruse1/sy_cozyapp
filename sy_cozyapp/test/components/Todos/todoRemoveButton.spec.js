'use strict'

/* eslint-env jest */

import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

// we import the not connected component version for testing
import { TodoRemoveButton } from 'components/Todos/TodoRemoveButton'
import Button from 'cozy-ui/react/Button'

// mock data for testing
import mockTodos from './_mocksTodos.json'

Enzyme.configure({ adapter: new Adapter() })

// async mock function
const mockDeleteTodo = jest.fn(() => Promise.resolve())
const todo = mockTodos[0]

describe('TodoRemoveButton component:', () => {
  it('should be rendered correctly', () => {
    const component = shallow(
      <TodoRemoveButton todo={todo} deleteTodo={mockDeleteTodo} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should handle isWorking correctly (display spinner)', () => {
    const component = shallow(
      <TodoRemoveButton todo={todo} deleteTodo={mockDeleteTodo} />
    )
    expect(component.state().isWorking).toBe(false)
    component.setState({ isWorking: true })
    expect(component).toMatchSnapshot()
  })

  it('should handle removeTodo correctly on button click', async () => {
    const component = shallow(
      <TodoRemoveButton todo={todo} deleteTodo={mockDeleteTodo} />
    )
    const removeButton = component.find(Button)
    await removeButton.simulate('click')
    expect(mockDeleteTodo.mock.calls.length).toBe(1)
    expect(mockDeleteTodo.mock.calls[0][0]).toBe(todo)
  })
})
