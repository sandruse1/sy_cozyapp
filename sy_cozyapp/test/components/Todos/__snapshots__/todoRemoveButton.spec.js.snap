// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoRemoveButton component: should be rendered correctly 1`] = `
<DefaultButton
  busy={false}
  className="todo-remove-button"
  disabled={false}
  extension="narrow"
  icon="delete"
  onClick={[Function]}
  theme="danger"
/>
`;

exports[`TodoRemoveButton component: should handle isWorking correctly (display spinner) 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoRemoveButton
    deleteTodo={[MockFunction]}
    todo={
      Object {
        "_id": 1234,
        "name": "Todo 1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "busy": true,
      "className": "todo-remove-button",
      "disabled": true,
      "extension": "narrow",
      "icon": "delete",
      "onClick": [Function],
      "theme": "danger",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "busy": true,
        "className": "todo-remove-button",
        "disabled": true,
        "extension": "narrow",
        "icon": "delete",
        "onClick": [Function],
        "theme": "danger",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
